### DISCUSSION TITLE
Repository, DAO의 차이와 패키지 구조

### DISCUSSION THREAD

**Discussion by GIVEN53** [2024-woowacourse-study/level-interview#123] [Spring 뒷고기]
.

**Comment by le2sky** [comment]
하나의 객체가 테이블 하나를 담당하는 것을 테이블 데이터 게이트웨이 패턴이라고 합니다. 
도메인 레이어를 트랜잭션 스크립트(절차지향적)로 구성하는 경우에 주로 사용됩니다. 
J2EE Core 패턴에서 이를 DAO(Data Access Object)라고 하며, 단일 테이블 또는 뷰에 접속하는 SQL을 처리합니다.
DAO는 후술할 Repository보다 DB와 더 연관있습니다.

Repository는 DDD에서 소개되었습니다. (Repository 패턴)
도메인 객체를 지속적으로 사용하려면 저장소에 보관해야합니다. 이를 위한 도메인 모델을 Repository라고 합니다.
Repository는 애그리거트 단위로 도메인 객체를 저장하고 조회하는 기능을 정의합니다.

Repository 인터페이스는 실제 저장소가 무엇인지와 관계없이, 도메인 객체에 대한 저장소를 위한 계약을 정의합니다.
도메인 모델의 관점에서 Repository 인터페이스는 도메인 객체를 영속화하는 데 필요한 기능을 추상화했기 때문에 고수준 모듈에 속합니다. 반면, Repository의 구현체는 저수준 모듈에 속하게 됩니다.(인프라 영역)
따라서, Repository 인터페이스는 domain 패키지에 Repository 구현체는 infra 패키지 혹은 그 이외에 존재합니다. ex) spring-data-jpa

이렇게 구현과 계약을 분리하는 것을 분리-인터페이스 패턴이라고 합니다. 
JDBC 관련 인터페이스는 java.sql 패키지에서 제공하고, 실제 구현체는 의존성에 추가된 드라이버 내부 패키지에 존재하는 것을 떠올려 볼 수 있습니다. 만약 JDBC 관련 인터페이스가 java.sql이 아닌, 구체적인 드라이버 내부에 존재했다면, 유연하게 다른 드라이버로 변경하기 어려울 수 있습니다.

Repository에서 관리하는 애그리거트 단위는 하나의 테이블 이상일 수 있으며, 테이블이 아닐 수도 있습니다.
이처럼 Repository는 도메인 모델에 더욱 초점을 맞추는 것이 DAO와의 차이점이라고 생각합니다.

https://www.youtube.com/watch?v=26S4VFUWlJM&ab_channel=WorkerKSUG
https://ttl-blog.tistory.com/1285
https://hudi.blog/separated-interface-pattern/
최범균님 DDD
마틴 파울러 블로그
J2EE Core Pattern

**Comment by J-I-H-O** [comment]
DAO(Data Access Object)는 데이터에 접근하기 위한 객체입니다. 
데이터 접근 로직을 저수준으로 추상화하여 DB 연결 관리나 쿼리를 포함할 수 있습니다.
데이터 소스에 대한 작업을 캡슐화 한 것으로, 그 자체가 영속성의 추상화이기 때문에 도메인 계층이 아닌 Data Access Layer에 속합니다.

Repository는 객체 Collection의 관리를 추상화한 매커니즘으로, Collection과 유사한 인터페이스를 제공해 도메인과 Data Access Layer 사이를 중재합니다.
객체의 Collection을 다루는 듯한 인터페이스를 제공한다는 점에서 DB 테이블보다 객체 중심의 Layer라고 할 수 있습니다. 즉 도메인과 더 가까운 높은 수준에 있으므로, Repository의 인터페이스는 도메인 계층에 속해야 합니다.
데이터 접근 로직을 더 높은 수준으로 추상화하여, 다양한 데이터 소스(데이터베이스, 파일 등)와 상호작용할 수 있습니다.

DAO는 데이터베이스의 CRUD 쿼리와 1:1 매칭되는 세밀한 단위의 작업을 제공합니다.
반면 Repository는 메모리에 로드된 객체 Collection에 대한 집합 처리를 위한 인터페이스를 제공합니다. 
결과적으로 Repository가 제공하는 하나의 작업이 DAO의 여러 작업을 호출하는 것이 일반적입니다.

```java
public class UserRepositoryImpl implements UserRepository {
    private UserDao userDao;    // user 테이블과 상호작용하는 DAO
    private ReservationDao reservationDao;  // reservation 테이블과 상호작용하는 DAO
    
    @Override
    public User get(Long id) {
        User user = (User) userDao.read(id); // UserDao를 사용해 user 정보 조회
        List<Reservation> reservations = reservationDao.findReservationsByEmail(user.getEmail());  // ReservationDao를 사용해 user의 예약 목록 조회
        user.setReservations(reservations);
        
        return user;
    }
}
```

https://isaac56.github.io/etc/2021/08/29/difference_DAO_Repository/
https://www.baeldung.com/java-dao-vs-repository
